### PR TITLE
Clocking api integration

### PR DESCRIPTION
**Pull request by harishsaini**
The Pull request takes care of everything on the right side of clockin clockout view controller.
It just needs the userjob object to be set or bind to something before presenting and it will do the rest of tasks, clockin, clock out and change state accordingly.

### CODE REVIEW

**Comment by ahujamanish** [comment]
@harishsaini always better to add some description to your pull requests to describe what exactly have you implemented.

File: Avingo/Code/Helpers/Extensions/String+Extensions.swift [hunk 1]
```diff
@@ -19,4 +20,18 @@ extension String {
 		let nonEmptyList = stringList.compactMap { $0 }
 		return nonEmptyList.joined(separator: separator)
 	}
+
+	func withHighlightedText(highlightedPartsOfString: [String], font: UIFont, color: UIColor, highlightFont: UIFont, highlightedTextColor: UIColor? = nil, searchHighlightedPartsOfStringFromLast: Bool = false) -> NSAttributedString {
```
Comment (tinman-fueled) [inline Avingo/Code/Helpers/Extensions/String+Extensions.swift]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Fueled/avingo-ios/blob/ccb759c3f5eff162d8c2c76839b390300a06d40c/Avingo/Code/Helpers/Extensions/String+Extensions.swift#L24"></span>Line Length Violation: Line should be 220 characters or less: currently 230 characters (line_length)</td>
    </tr>
  </tbody>
</table>

File: Avingo/Code/UI/ClockIn/ClockInViewController.swift [hunk 2]
```diff
@@ -31,15 +24,36 @@ final class ClockInViewController: UIViewController {
 		super.viewDidLoad()
 		self.setupBindings()
 	}
-
+
```
Comment (tinman-fueled) [inline Avingo/Code/UI/ClockIn/ClockInViewController.swift]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Fueled/avingo-ios/blob/ccb759c3f5eff162d8c2c76839b390300a06d40c/Avingo/Code/UI/ClockIn/ClockInViewController.swift#L27"></span>Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)</td>
    </tr>
  </tbody>
</table>

File: Avingo/Code/UI/ClockIn/ClockInViewController.swift [hunk 3]
```diff
@@ -31,15 +24,36 @@ final class ClockInViewController: UIViewController {
 		super.viewDidLoad()
 		self.setupBindings()
 	}
-
+	
 	private func setupBindings() {
```
Comment (tinman-fueled) [inline Avingo/Code/UI/ClockIn/ClockInViewController.swift]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Fueled/avingo-ios/blob/ccb759c3f5eff162d8c2c76839b390300a06d40c/Avingo/Code/UI/ClockIn/ClockInViewController.swift#L28"></span><code>instance method setupBindings()</code> takes 241.04 ms to build</td>
    </tr>
  </tbody>
</table>

**Comment by harishsaini** [comment]
The Pull request takes care of everything on the right side of clockin clockout view controller. 
It just needs the user userjob object to be set or bind to something before presenting and it will do the rest of tasks, clockin, clock out and change state accordingly. 

Can't find a way to add description now 
@ahujamanish

File: Avingo/Code/UI/ClockIn/ClockInViewController.swift [hunk 4]
```diff
@@ -35,11 +28,32 @@ final class ClockInViewController: UIViewController {
...
```
Comment (tinman-fueled) [inline Avingo/Code/UI/ClockIn/ClockInViewController.swift]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Fueled/avingo-ios/blob/58bb1aa499f2e30f1e5f1bc60e9d80a295d675e0/Avingo/Code/UI/ClockIn/ClockInViewController.swift#L28"></span><code>instance method setupBindings()</code> takes 233.54 ms to build</td>
    </tr>
  </tbody>
</table>

File: Avingo/Code/Helpers/Classes/UserSettingsManager.swift [hunk 5]
```diff
@@ -13,20 +13,29 @@ import KeychainAccess
...
 	let isUserLoggedIn: Property<Bool>
 
-	private let coreDataUser = MutableProperty<User?>(nil)
-	private let logout = Signal<Void, NoError>.pipe()
+	// This will return a unique device id until the device is reset
+	lazy var uniqueDeviceId: String = {
+		if let deviceId = keychain[UserSettingsManager.keychainDeviceIdKey] {
+			return deviceId
+		} else {
+			let deviceId = UIDevice.current.identifierForVendor?.uuidString ?? "Invalid Device Id"
```
[the remark this answers is not in the document]
Comment (harishsaini) [reply Avingo/Code/Helpers/Classes/UserSettingsManager.swift]: @ahujamanish Yes, i was curious to know the case when we actually don't get any id, but that that can be identified by timestamp also, fixed.

File: Avingo/Code/Helpers/Extensions/Formatter+Avingo.swift [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+//  Formatter+Avingo.swift
+//  Avingo
+//
+//  Created by Harish Saini on 08/06/18.
+//  Copyright © 2018 Fueled. All rights reserved.
+//
+
+import Foundation
+
+extension Formatter {
```
[the remark this answers is not in the document]
Comment (harishsaini) [reply Avingo/Code/Helpers/Extensions/Formatter+Avingo.swift]: yes, that's a mistake, thanks

File: Avingo/Code/UI/ClockIn/ClockInViewController.swift [hunk 7]
```diff
@@ -35,11 +28,32 @@ final class ClockInViewController: UIViewController {
...
 		self.titleLabel.reactive.attributedText <~ self.viewModel.title
 		self.subTitleLabel.reactive.attributedText <~ self.viewModel.subTitle
-		self.descriptionLabel.reactive.attributedText <~ self.viewModel.description
+		self.descriptionLabel.reactive.attributedText <~ self.viewModel.descriptionText
```
[the remark this answers is not in the document]
Comment (harishsaini) [reply Avingo/Code/UI/ClockIn/ClockInViewController.swift]: fixed, thanks for `ReactiveExtensionsProvider`

File: Avingo/Code/UI/ClockIn/ClockInViewController.swift [hunk 8]
```diff
@@ -35,11 +28,32 @@ final class ClockInViewController: UIViewController {
...
+			self.viewModel.clockOutAction.errors
+			])
+		self.reactive.makeBindingTarget { (controller, error: AnyError) in
+			controller.presentAlert(message: error.errorMessage)
+		} <~ mergedErrors
 	}
 
-	@IBAction private func clockinClockOutButtonTapped() {
+	private func dismissAfterDelay() {
+		DispatchQueue.main.asyncAfter(deadline: .now() + 15) { [weak self] in
```
[the remark this answers is not in the document]
Comment (harishsaini) [reply Avingo/Code/UI/ClockIn/ClockInViewController.swift]: sure

**Comment by harishsaini** [comment]
@ahujamanish made the changes

File: Avingo/Code/Helpers/Extensions/String+Extensions.swift [hunk 9]
```diff
@@ -19,4 +20,20 @@ extension String {
 		let nonEmptyList = stringList.compactMap { $0 }
 		return nonEmptyList.joined(separator: separator)
 	}
+
+	func withHighlightedText(highlightedPartsOfString: [String], font: UIFont, color: UIColor,
+							 highlightFont: UIFont, highlightedTextColor: UIColor? = nil,
```
Comment (tinman-fueled) [inline Avingo/Code/Helpers/Extensions/String+Extensions.swift]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Fueled/avingo-ios/blob/f9295004712c3928ca2a9186e1d426bd79b3a3b2/Avingo/Code/Helpers/Extensions/String+Extensions.swift#L25"></span>Vertical Parameter Alignment Violation: Function parameters should be aligned vertically if they’re in multiple lines in a declaration. (vertical_parameter_alignment)</td>
    </tr>
  </tbody>
</table>

File: Avingo/Code/Helpers/Extensions/String+Extensions.swift [hunk 10]
```diff
@@ -19,4 +20,20 @@ extension String {
...
+							 searchHighlightedPartsOfStringFromLast: Bool = false) -> NSAttributedString {
```
Comment (tinman-fueled) [inline Avingo/Code/Helpers/Extensions/String+Extensions.swift]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Fueled/avingo-ios/blob/f9295004712c3928ca2a9186e1d426bd79b3a3b2/Avingo/Code/Helpers/Extensions/String+Extensions.swift#L26"></span>Vertical Parameter Alignment Violation: Function parameters should be aligned vertically if they’re in multiple lines in a declaration. (vertical_parameter_alignment)</td>
    </tr>
  </tbody>
</table>

**Comment by tinman-fueled** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/Fueled/avingo-ios/blob/f9295004712c3928ca2a9186e1d426bd79b3a3b2/Avingo/Code/UI/ClockIn/ClockInViewController.swift#L29">Avingo/Code/UI/ClockIn/ClockInViewController.swift#L29</a> - <code>instance method setupBindings()</code> takes 243.28 ms to build</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Pull request merged**